project/data_analysis/python/leakage/get_leakage_covariance.py

```diff
@@ -0,0 +1,141 @@
+
+if d["cov_T_E_only"] == True:
+    modes_for_cov = ["TT", "TE", "ET", "EE"]
+else:
+    modes_for_cov = spectra
+
+pspy_utils.create_directory(plot_dir)
+
+# read the leakage model
+gamma, err_m, var = {}, {}, {}
+name_list = []
+
+plt.figure(figsize=(12, 8))
+for sv in surveys:
+    arrays = d[f"arrays_{sv}"]
+    for ar in arrays:
+    
+        name = f"{sv}_{ar}"
+        
+        log.info(f"reading leakage info {name}")
+
+        gamma[name], err_m[name], var[name] = {}, {}, {}
+        l, gamma[name]["TE"], err_m[name]["TE"], gamma[name]["TB"], err_m[name]["TB"] = leakage.read_leakage_model(leakage_file_dir,
+                                                                                                                   d[f"leakage_beam_{name}"],
+                                                                                                                   lmax,
+                                                                                                                   lmin=2)
+
+        cov = {}
+        cov["TETE"] = leakage.error_modes_to_cov(err_m[name]["TE"])
+        cov["TBTB"] = leakage.error_modes_to_cov(err_m[name]["TB"])
+
+        var[name]["TETE"] = cov["TETE"].diagonal()
+        var[name]["TBTB"] = cov["TBTB"].diagonal()
+        var[name]["TETB"] = var[name]["TETE"] * 0
+        
+        for field in ["TE", "TB"]:
+            corr = so_cov.cov2corr(cov[field + field])
+
+            plt.figure(figsize=(12, 8))
+            plt.imshow(corr)
+            plt.title(f"gamma {field} correlation {name}", fontsize=12)
+            plt.colorbar()
+            plt.savefig(f"{plot_dir}/gamma_{field}_correlation_{name}.png", bbox_inches="tight")
+            plt.clf()
+            plt.close()
+        
+        name_list += [name]
+
+spec_name_list = pspipe_list.get_spec_name_list(d, char="_")
+
+ps_th_dict, residual = {}, {}
+for spec_name in spec_name_list:
+
+    name1, name2 = spec_name.split("x")
+    
+    l_th, ps_th_dict[spec_name] = so_spectra.read_ps(f"{bestfit_dir}/cmb_and_fg_{spec_name}.dat", spectra=spectra)
+    id = np.where(l_th < lmax)
+    
+    l_th = l_th[id]
+    for spec in spectra:
+        ps_th_dict[spec_name][spec] = ps_th_dict[spec_name][spec][id]
+
+    l, residual[spec_name] = leakage.leakage_correction(l_th,
+                                                        ps_th_dict[spec_name],
+                                                        gamma[name1],
+                                                        var[name1],
+                                                        lmax,
+                                                        return_residual=True,
+                                                        gamma_beta=gamma[name2],
+                                                        binning_file=binning_file)
+
+
+ps_all_corrected = {}
+
+for iii in range(n_sims):
+
+    log.info(f"sims {iii}/{n_sims}")
+
+    gamma_sim = {}
+    for name in name_list:
```

Comment:
I think `name_list` is actually not needed. It is only used here and the `for` loop can be changed by something like

```python
for name in gamma_sim.keys():
  ...
```